### PR TITLE
smb: add smb.version keyword

### PR DESCRIPTION
**Pull request by jmtaylor90**
Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://docs.suricata.io/en/latest/devguide/codebase/contributing/contribution-process.html
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata.io/about/contribution-agreement/
- [x] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:

https://redmine.openinfosecfoundation.org/issues/5075

Describe changes:
- adds smb.version keyword
- adds smb.version documentation
-

### Provide values to any of the below to override the defaults.

To use a pull request use a branch name like `pr/N` where `N` is the
pull request number.

Alternatively, `SV_BRANCH` may also be a link to an
OISF/suricata-verify pull-request.

```
SV_BRANCH=pr/1366
```

### CODE REVIEW

**Comment by catenacyber** [comment]
Did you start from https://github.com/OISF/suricata/pull/7336 ?

**Comment by catenacyber** [comment]
Check rust failure seems unrelated : fix may be https://github.com/OISF/suricata/pull/9416/commits

**Comment by jmtaylor90** [comment]
> Did you start from #7336 ?

I did, sorry about that, I was going to reference/indicate that and forgot.

**Comment by jmtaylor90** [comment]
> > > Did you start from #7336 ?
> > 
> > 
> > I did, sorry about that, I was going to reference/indicate that and forgot.
> 
> Thanks for picking this back up. What did you change ? (just rebase ?)

This was largely a rebase and formatting (clang/clippy/rustfmt) changes. I did add some debug logging as well.

I did change some code to address an initial question you had around:
https://github.com/OISF/suricata/pull/7336/files#diff-f037023d6eda2c0a79e08ad1701daa9cd8842be75ba08c267990edc50d51926fR98

I removed that bit of code and updated the DetectAppLayerInspectEngineRegister2 calls.

I also added an additional signature validation check:
https://github.com/OISF/suricata/pull/9415/files#diff-f037023d6eda2c0a79e08ad1701daa9cd8842be75ba08c267990edc50d51926fR90

which is verified in the associated suricata verify  PR

File: src/detect-smb-version.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ */
+
+#ifndef __DETECT_SMB_VERSION_H__
+#define __DETECT_SMB_VERSION_H__
+
+/** \brief registers the keyword into the engine. Called from
+ *         detect.c::SigTableSetup() */
+void DetectSmbVersionRegister(void);
+
+#endif /* __DETECT_SMB-VERSION_H__ */
```
Comment (catenacyber) [inline src/detect-smb-version.h:25]: typo - _

File: src/detect-smb-version.c [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+#include "detect-engine-state.h"
+#include "detect-engine-prefilter.h"
+#include "detect-engine-content-inspection.h"
+
+#include "detect-smb-version.h"
+#include "rust.h"
+
+#define BUFFER_NAME         "smb_version"
+#define KEYWORD_NAME        "smb.version"
+#define KEYWORD_NAME_LEGACY BUFFER_NAME
```
Comment (catenacyber) [inline src/detect-smb-version.c:43]: no legacy needed

**Review by catenacyber — CHANGES_REQUESTED** [review]
Thanks for the changes

- CI : ✅  rust failure is unrelated
- Code : looks good, some nits to review
- Commits segmentation : maybe the rustfmt can be its own first commit
- Commit messages : Could you put the ticket number in the commit message please ? 
- Git ID set : looks fine for me
- CLA : I do not have access, but looks like some commits were already merged
- Doc update : really nice  👏 
- Redmine ticket : ok
- Rustfmt : ok
- Tests : Suricata-verify tests look fine
- Dependencies added: none

File: src/detect-smb-version.h (hunk 1 above)
Comment (jmtaylor90) [reply src/detect-smb-version.h:25]: fixed in next PR

File: src/detect-smb-version.c (hunk 2 above)
Comment (jmtaylor90) [reply src/detect-smb-version.c:43]: fixed in next PR

File: src/detect-smb-version.c [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+{
+    SCEnter();
+
+    SCLogDebug("smb_version: DetectSmbVersionMatchRust");
+
+    if (f->alproto == ALPROTO_SMB) {
+        return rs_smb_version_match(txv, (void *)m);
+    }
+
+    if (rs_smb_version_match(txv, (void *)m) != 1)
```
[the remark this answers is not in the document]
Comment (jmtaylor90) [reply src/detect-smb-version.c:136]: I am not sure, I saw this sort of if block pattern in a couple other places and figured it came in here in matching style. Example : https://github.com/OISF/suricata/blob/master/src/detect-dce-opnum.c#L105

I also came across:
https://github.com/OISF/suricata/blob/master/src/detect-dce-iface.c#L114

Where there is some logging added. Looking at this again, I kind of like the additional logging but of course defer to what you think the best route to go is here.

Additional logging wise I was thinking something like:

```
if (f->alproto == ALPROTO_SMB) {
        return rs_smb_version_match(txv, (void *)m);
    }

    if (rs_smb_version_match(txv, (void *)m) != 1) {
        SCLogDebug("rs_smb_version_match: didn't match");
        SCReturnInt(0);
    } else {
        SCLogDebug("rs_smb_version_match: matched!");
    }

    SCReturnInt(1);
}
```

I think ALPROTO_SMB is the right one to use here for obtaining version information.

[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-smb-version.c:136]: So, I think DCERPC keywords can work for both SMB and DCERPC, but SMB keywords are SMB only, as enforced in the setup function.
So you need to call only once `rs_smb_version_match` and not check for `f->alproto == ALPROTO_SMB`

[the remark this answers is not in the document]
Comment (jmtaylor90) [reply src/detect-smb-version.c:136]: Ah okay, I see. I will see about removing that check. What do you think with regards to adding the logging block?

File: rust/src/smb/detect.rs [hunk 4]
```diff
@@ -156,11 +152,9 @@ pub extern "C" fn rs_smb_tx_get_dce_iface(state: &mut SMBState,
...
 
 #[no_mangle]
-pub unsafe extern "C" fn rs_smb_tx_get_ntlmssp_user(tx: &mut SMBTransaction,
-                                            buffer: *mut *const u8,
-                                            buffer_len: *mut u32)
-                                            -> u8
-{
+pub unsafe extern "C" fn rs_smb_tx_get_ntlmssp_user(
+    tx: &mut SMBTransaction, buffer: *mut *const u8, buffer_len: *mut u32,
+) -> u8 {
```
Comment (jufajardini) [inline rust/src/smb/detect.rs:157]: Are these purely formatting changes to existing code? If yes, we usually prefer to see those in a separate commit, so its easier to visualize and review :)

Comment (jmtaylor90) [reply rust/src/smb/detect.rs:157]: They are, I will have the rustfmt/clippy changes in a separate commit in the next PR.

Comment (jufajardini) [reply rust/src/smb/detect.rs:157]: Happens to us all, thanks! :)

File: doc/userguide/rules/smb-keywords.rst [hunk 5]
```diff
@@ -58,3 +58,45 @@ Examples::
...
+
+On the other hand SMB3 is more an extension for SMB2. When using SMB2 we can select one of the following dialects for the conversation between client and server:
+
+- 2.0.2
+- 2.1
+- 3.0
+- 3.0.2
+- 3.1.1
+
+We say we are using SMB3 when we select a 3.x dialect for the conversation, so you can use SMB3.0, SMB3.0.2 or SMB3.1.1. The higher you choose, the more capabilities you have, but the message syntax and message command number remains the same.
```
Comment (jufajardini) [inline doc/userguide/rules/smb-keywords.rst:102]: While I liked the more in-depth explanation here, it made me a bit confused as to how this impacts the keyword usage, if at all. 

Can you add a note on this, or maybe complement it with rule examples, if this is something that is relevant to rule writers? :)

Comment (jmtaylor90) [reply doc/userguide/rules/smb-keywords.rst:102]: Sure, will revisit for the next PR and update the -verify tests as needed

File: rust/src/smb/detect.rs [hunk 6]
```diff
@@ -1,4 +1,4 @@
-/* Copyright (C) 2017 Open Information Security Foundation
+/* Copyright (C) 2023 Open Information Security Foundation
```
Comment (jufajardini) [inline rust/src/smb/detect.rs:1]: Copyright years are updated in the form of: initial year-last updated year
See https://github.com/OISF/suricata/pull/9415/commits/482eb11213677d4f2b56e52757b5810b112cd67d#diff-809d44481dd655fd192c9767d520c5ced78d5c1de6f2b94ee4255e91c1376d53L1 ;)

Comment (jmtaylor90) [reply rust/src/smb/detect.rs:1]: updated in next PR

**Comment by jmtaylor90** [comment]
continued in #9427

**Pull request closed without merging**